Update changelog for Themis 0.12.0 (#531)
cossacklabs/themis: CHANGELOG.md

# Themis ChangeLog

## [0.12.0](https://github.com/cossacklabs/themis/releases/tag/0.12.0), September 23rd 2019

**TL;DR:**
Added WasmThemis to support WebAssembly,
improved packaging and platform compatibility,
additional safety checks and tricky bug fixes.

**Breaking changes:**

> - Default installation path has been changed from `/usr` to `/usr/local`.

Author: On macOS it has been installed to `/usr/local` since forever. I guess we can say ‘on Linux’ then.

- Some API of GoThemis have been renamed (old API is considered deprecated).
- Deprecated `rubythemis` gem has been completely removed.

_Code:_

- **Core**

  - **Soter** (low-level security core used by Themis)

    - New function: `soter_wipe()`.
      It can be used to securely wipe sensitive data from memory after it's no longer necessary.
      All Themis cryptosystems now make use of this new API.
      ([#488](https://github.com/cossacklabs/themis/pull/488))

>     - Non-cryptographically secure PRNG could be used if the backend does not provide a cryptographically secure one.

Reviewer: ```suggestion
    - Improved usage and error handling of OpenSSL/BoringSSL PRNGs.
```

actually, i don't see how this commit solves "Non-cryptographically secure PRNG could be used if ..."

Author: `RAND_bytes()` could return negative value to indicate that CSPRNG is not available and leave the buffer in undefined state (no idea what OpenSSL actually does there). We were treating this error code as success and went on with our business, assuming that we got the random bytes we needed. With this change we correctly report an error and do not assume that ‘random crap found on the call stack’ is a PRNG in any capacity.

Anyway, I like your wording better for being concise and significant.

      Thanks to [**@veorq**](https://github.com/veorq) for finding this issue.
      ([#485](https://github.com/cossacklabs/themis/pull/485))

    - Multiple memory safety fixes in OpenSSL/BoringSSL usage.
      Thanks to [**@outspace**](https://github.com/outspace) for identifying these issues.
      ([#501](https://github.com/cossacklabs/themis/pull/501),
       [#524](https://github.com/cossacklabs/themis/pull/524),
       [#525](https://github.com/cossacklabs/themis/pull/525))

  - **Key generation**

    - Improved key validity checks across all Themis cryptosystems.
      ([#486](https://github.com/cossacklabs/themis/pull/486))

    - EC key generator could produce an invalid public key without returning an error.
      Thanks to [**@vixentael**](https://github.com/vixentael) for finding this issue.
      ([#500](https://github.com/cossacklabs/themis/pull/500))

  - **Secure Cell**

    - Improved handling of edge cases with large data buffers.
      ([#496](https://github.com/cossacklabs/themis/pull/496),
       [#497](https://github.com/cossacklabs/themis/pull/497))

>     - Minor performance improvements in encryption/decryption code path.

Reviewer: ```suggestion
    - Improved performance of SecureCell's encryption/decryption procedures.
```

      ([#496](https://github.com/cossacklabs/themis/pull/496))

  - **Library ABI**

    - Themis shared libraries now have a formal ABI version.
      This ensures that no compatibility issues arise if we ever need to introduce breaking changes in ABI.
      ([#454](https://github.com/cossacklabs/themis/pull/454))

>     - Remove private symbols from public export lists.

Reviewer: ```suggestion
    - Removed private symbols from public export lists.
```

       [#472](https://github.com/cossacklabs/themis/pull/472))

  - <a id="0.12.0-packaging-updates">**Installation & packaging**</a>

    - Themis is now installed to `/usr/local` by default when building from source.
      ([#448](https://github.com/cossacklabs/themis/pull/448))

      This _may_ be a **breaking change** if you system has non-standard precedence rules.
      If you install Themis from source code directly, please upgrade cleanly like this:

      ```bash
      make uninstall PREFIX=/usr
      make install   PREFIX=/usr/local
      ```

      Please consider using [binary repositories](https://docs.cossacklabs.com/pages/documentation-themis/#installing-themis-from-repositories) to install Themis.
      If your system or package manager is not supported yet, please let us know via dev@cossacklabs.com.

    - Themis packages now support multiarch installations.
      ([#512](https://github.com/cossacklabs/themis/pull/512))

      Multiarch enables parallel installation of 32-bit and 64-bit versions of the library.
      This is particularly important on CentOS where some tools like _pkg-config_ failed to locate Themis due to non-standard installation path.

>     - Makefile now supports a number of [standard GNU variables](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html) like `libdir`.

Reviewer: ```suggestion
    - Updated Makefile to support a number of [standard GNU variables](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html) like `libdir`.
```

      ([#453](https://github.com/cossacklabs/themis/pull/453),
       [#455](https://github.com/cossacklabs/themis/pull/455))

>     - Accurate dependencies for development packages make sure you don't have to install anything manually after installing Themis from package repositories.

Reviewer: ```suggestion
    - Made more accurate dependencies for development packages make sure you don't have to install anything manually after installing Themis from package repositories.
```

      ([#446](https://github.com/cossacklabs/themis/pull/446))

>     - NSIS installer is now available for Windows.

Reviewer: add smth about this in `TLDR` section? it might be a large change for some of our users

      | Old API                                | New API                    |
      | -------------------------------------- | -------------------------- |
      | `cell.CELL_MODE_SEAL`                  | `cell.ModeSeal`            |
      | `compare.COMPARE_MATCH`                | `compare.Match`            |
      | `keys.KEYTYPE_EC`                      | `keys.TypeEC`              |
      | `session.STATE_ESTABLISHED`            | `session.StateEstablished` |
      | `(*session.SecureSession) GetRemoteId` | `GetRemoteID`              |

  - GoThemis is compatible with Go 1.11 modules since this release.
    For example, you are now able to pin a specific version of GoThemis in your projects.
    ([#505](https://github.com/cossacklabs/themis/pull/505))

- **iOS, macOS**

  - Example code and projects for Objective-C and Swift are now up-to-date, cleaned up, and modernized.
    ([#463](https://github.com/cossacklabs/themis/pull/463),
     [#467](https://github.com/cossacklabs/themis/pull/467))

  - OpenSSL version is pinned to 1.0.2.17 for Carthage package to work around a compilation issue with the latest version.
    ([#484](https://github.com/cossacklabs/themis/pull/484))

>     <!-- TODO: Isn't this scheduled for a fix in 0.12? -->

Reviewer: it is, if new OpenSSL would work well.
